Fixed unhelpful error message in from_thread_run functions.

guilledk opened the pull request:
Added helpful error message in the case the user passed incorrect type of function to both "from_thread_run" functions.

Closes #1244

guilledk commented:
I'm not really familiar with the whole CI thing, so just point me to some docs and I'll fix the test fails.

Didn't expect a simple error message would break so many tests :P

oremanj commented on trio/_threads.py:
```diff
@@ -377,6 +379,9 @@ def from_thread_run(afn, *args, trio_token=None):
           "foreign" thread, spawned using some other framework, and still want
           to enter Trio.
     """
+    if not inspect.iscoroutinefunction(afn):
```
Unfortunately, identifying async functions is rather subtler than just `inspect.iscoroutinefunction`, because async functions can be wrapped in a way that doesn't satisfy that test. (For example, Trio functions decorated with `@enable_ki_protection` are implemented using a wrapper that looks synchronous. `functools.partial` objects don't even satisfy `inspect.isfunction`, let alone `inspect.iscoroutinefunction`. All of these are perfectly valid to pass as an "async function" as long as the function being wrapped is async.)

The best way to see if something is an async function is to call it without `await` and use `inspect.iscoroutine` on the result. Of course, if it wasn't an async function, you just executed it, so it's best to do this only in the context where you were planning to call it anyway. In this case, that would be inside `unprotected_afn` below.

If you're not willing to call the function, you have to get a lot more special-cased to see if it's async or not. You can look inside the `.func` attribute of a `functools.partial` object, follow the `.__wrapped__` attributes generally added by function wrappers, etc, but there's no approach that will definitely succeed for any async callable.

oremanj commented on trio/_threads.py:
```diff
@@ -422,6 +427,9 @@ def from_thread_run_sync(fn, *args, trio_token=None):
           "foreign" thread, spawned using some other framework, and still want
           to enter Trio.
     """
+    if not (inspect.isfunction(fn) and not inspect.iscoroutinefunction(fn)):
```
As discussed above, these won't tell you what you want. Better to wait and raise an error if the result of calling the function satisfies `inspect.iscoroutine`. (If it does, you'll also want to `close()` it before returning, so you don't get an extra "coroutine was never awaited" warning.)

guilledk replied:
Just saw your comment on #1504, should I look into `Runner.spawn_impl()` for the "async fn or error" logic?

oremanj replied:
That would be a useful extension if you feel like it! The relevant code is toward the top of spawn_impl(), with the comments about "Give good error for [...]"

guilledk commented on trio/_threads.py:
```diff
@@ -377,6 +379,9 @@ def from_thread_run(afn, *args, trio_token=None):
           to enter Trio.
     """
+    if not inspect.iscoroutinefunction(afn):
+        raise AttributeError("afn must be an asynchronous function")
```
Great! I knew the error message needed some work.

guilledk commented:
Thanks a lot for the feedback!, I'll work on it and send another PR!

njsmith commented:
@guilledk Note that you don't have to actually create a new PR from scratch – you can just more commits to your `issue1244` branch on github, and they'll automatically appear here in this PR.

guilledk commented:
Ok good to know.

Thanks a lot for the patience, you guys rock!

guilledk commented:
So following what @oremanj said in #1504, I moved the "either get a coroutine from this async function+args or raise an error explaining why it's not in the form I wanted" logic to a function in `trio._util` called `coroutine_or_error` (WIP name).

Added the check to `trio._threads.from_thread_run`, but its still missing the check on `trio._threads.from_thread_run_sync`, I'll do that one tomorrow.

Also there is some code formatting issues I have to fix.

guilledk commented:
Cool, all that's missing are the tests!

guilledk commented:
@njsmith Hey so I could use some pointers to fix the code coverage issues, I check the codecov diff and I don't seem to get were the issue is. Thanks in advance!

pquentin commented:
The coverage drop is in trio/socket.py: https://codecov.io/gh/python-trio/trio/pull/1513/src/trio/socket.py

![Screenshot 2020-05-14 at 15 26 38](https://user-images.githubusercontent.com/42327/81929040-51cbe880-95f7-11ea-9189-cb3e40b92708.png)

The issue is that the macOS 3.6 build was cancelled after 1m30s, I don't know why: https://github.com/python-trio/trio/pull/1513/checks?check_run_id=672946337

Closing/reopening to re-run all jobs

guilledk commented:
Cool so this closes #1244

pquentin commented:
Great that the tests and coverage pass!

@guilledk Note that you need to edit your first message to say that, otherwise GitHub won't close the issue when we merge this.

guilledk commented:
@pquentin Done!

oremanj commented on trio/_core/_run.py:
```diff
@@ -1243,86 +1242,8 @@ def spawn_impl(self, async_fn, args, nursery, name, *, system_task=False):
             assert self.init_task is None
 
-        ######
-        # Call the function and get the coroutine object, while giving helpful
```
This is a better description than the "Check if async_fn is callable coroutine" that you added below, so I'd suggest you keep it. (The function is not a "callable coroutine" -- it's a callable that returns a coroutine object, but that's just another way of saying "async function" which is Trio's preferred term. Coroutine objects themselves are not callable.)

oremanj commented on trio/_threads.py:
```diff
@@ -404,12 +408,13 @@ def from_thread_run_sync(fn, *args, trio_token=None):
         RunFinishedError: if the corresponding call to `trio.run` has
             already completed.
         Cancelled: if the corresponding call to `trio.run` completes
```
Not your fault, but this is copypasta from the async version -- sync functions can't raise Cancelled. So you can just remove the description of Cancelled as an exception-that-might-be-raised entirely.

oremanj commented on trio/_threads.py:
```diff
@@ -404,12 +408,13 @@ def from_thread_run_sync(fn, *args, trio_token=None):
             which would otherwise cause a deadlock.
         AttributeError: if no ``trio_token`` was provided, and we can't infer
-            one from context.
+            one from context. Also if ``fn`` is not a sync function.
```
This is leftover from your previous use of AttributeError for this condition, but you're now using TypeError.

oremanj replied:
You marked this as resolved but you didn't update the docstring.

oremanj replied:
You still have not updated the docstring despite having marked this as resolved again. The request is to remove the text "Also if fn is not a sync function" from the description for AttributeError, because you no longer raise AttributeError in that case.

oremanj commented on trio/_threads.py:
```diff
@@ -422,6 +427,13 @@ def from_thread_run_sync(fn, *args, trio_token=None):
           to enter Trio.
     """
+
+    if not callable(fn) or inspect.iscoroutinefunction(fn):
```
I actually don't think you need to explicitly check whether it's callable, because the TypeError you get from calling a non-callable will properly propagate back into the caller of `trio.from_thread.run_sync` when you try to call the function below.

As I explained in chat, `inspect.iscoroutinefunction` will not catch all async callables. That's OK here, but you will give a better user experience (ie, catch more problem cases) if you wait until after you've called the function, and then throw an error if the value you got back from it satisfies `inspect.iscoroutine`.

oremanj commented on trio/_util.py:
```diff
@@ -85,6 +88,87 @@ def is_main_thread():
+        # Give good error for: nursery.start_soon(func_returning_future)
+        if _return_value_looks_like_wrong_library(coro):
+            raise TypeError(
+                "start_soon got unexpected {!r} – are you trying to use a "
```
Suggest you change "start_soon" to "Trio" in this error and the one below. This is for consistency with the other error messages. start_soon was already incomplete in which cases it covered (you can get the same error from trio.run() or nursery.start()) but now it's even more so (since now from_thread.run(), and later potentially even more functions, will be using this helper too).

oremanj commented on trio/_util.py:
```diff
@@ -85,6 +88,87 @@ def is_main_thread():
+# Call the function and get the coroutine object, while giving helpful
+# errors for common mistakes. Returns coroutine object.
+######
+def coroutine_or_error(async_fn, args=[]):
```
Trio convention is to take `*args` in this case -- we almost never pass around the argument tuple as a tuple, even internally.

oremanj commented on trio/_threads.py:
```diff
@@ -422,6 +427,13 @@ def from_thread_run_sync(fn, *args, trio_token=None):
+    if not callable(fn) or inspect.iscoroutinefunction(fn):
+        raise TypeError(
+            "Trio expected a sync function, but {!r} appears to not be "
+            "callable or asynchronous".format(getattr(fn, "__qualname__", fn))
```
You know which it is (is it non-callable or is it an async function?), so the error message should say which it is.

Also, be extra careful about grammatical ambiguity in error messages, since we want them to be as comprehensible as possible. In this case "not be callable or asynchronous" could be read as "not be (callable or asynchronous)" where what you mean is "(not be callable) or (be asynchronous)". I would write that as "is not callable or is an async function".

njsmith replied:
Python already gives a decent error message if you try to call something that's non-callable, so we could drop that check entirely and just let Python raise the exception for us.

oremanj replied:
Oh, whoops, I thought I had left a comment about that but it seems I didn't.

I was also going to suggest that instead of `inspect.iscoroutinefunction` (which won't catch async methods, async functions used with `partial`, etc), you wait until you call the function below and then raise an error if the return value satisfies `inspect.iscoroutine`.

oremanj commented on trio/_core/tests/test_run.py:
```diff
@@ -1696,74 +1696,6 @@ async def test_current_effective_deadline(mock_clock):
-# @coroutine is deprecated since python 3.8, which is fine with us.
-@pytest.mark.filterwarnings("ignore:.*@coroutine.*:DeprecationWarning")
-def test_nice_error_on_bad_calls_to_run_or_spawn():
-    def bad_call_run(*args):
```
You've removed the test that passing these things to run() or start_soon() actually results in an error (ie, that your now factored-out helper actually gets called). You don't need to duplicate the full suite of tests, but you should still pass one or two error-provoking arguments and ensure that errors get produced.

guilledk replied:
Ok, added test for passing a sync function and an async generator.

oremanj commented on trio/tests/test_threads.py:
```diff
@@ -471,6 +471,15 @@ def thread_fn():
+    with pytest.raises(TypeError) as excinfo:
+        from_thread_run_sync(async_fn)
+
+    assert "expected a sync function" in str(excinfo.value)
```
It's nicer to write this as `with pytest.raises(TypeError, match="expected a sync function"):`

oremanj commented on trio/tests/test_threads.py:
```diff
@@ -488,6 +497,18 @@ def thread_fn():
+        from_thread_run(sync_fn)
+
+    with pytest.raises(TypeError) as excinfo:
+        await to_thread_run_sync(thread_fn)
```
You don't need the separate thread_fn. You can say `await to_thread_run_sync(from_thread_run, sync_fn)`

guilledk commented on trio/_core/tests/tutil.py:
```diff
@@ -52,6 +54,24 @@ def gc_collect_harder():
+# manager should be used anywhere this happens to hide those messages, because
+# when expected they're clutter.
+@contextmanager
+def ignore_coroutine_never_awaited_warnings():
```
Moved this here because its used in several tests now.

guilledk commented on trio/_threads.py:
```diff
@@ -425,7 +427,16 @@ def from_thread_run_sync(fn, *args, trio_token=None):
+            call = fn(*args)
+
+            if inspect.iscoroutine(call):
+                call.close()
```
To avoid `RuntimeWarning`s

oremanj replied:
Please say that in a comment in the code, so that people reading the code will understand it. Comments in the PR don't help anyone except the people reviewing the PR.

guilledk commented on trio/tests/test_threads.py:
```diff
@@ -13,7 +13,7 @@
 )
 
 from .._core.tests.test_ki import ki_self
-from .._core.tests.tutil import slow
```
Unused import

oremanj commented on trio/_core/_run.py:
```diff
@@ -1247,86 +1246,7 @@ def spawn_impl(self, async_fn, args, nursery, name, *, system_task=False):
-            )
-
-        ######
-        # Set up the Task object
```
This comment ("Set up the Task object") is still applicable to the code below, so you shouldn't remove it.

oremanj commented on trio/_threads.py:
```diff
@@ -425,7 +427,16 @@ def from_thread_run_sync(fn, *args, trio_token=None):
         @disable_ki_protection
         def unprotected_fn():
-            return fn(*args)
+            call = fn(*args)
```
This is the value returned by the function, so `call` is not a very clear name for it. `result` or `res` or `ret` perhaps?

oremanj commented on trio/_core/tests/test_run.py:
```diff
@@ -1709,59 +1693,25 @@ async def main():
 
-            async def f():  # pragma: no cover
-                pass
+        with ignore_coroutine_never_awaited_warnings():
```
You don't need this anymore since you added the logic to close the passed coroutine, right?

guilledk replied:
Yep! you're right

oremanj commented on trio/_core/tests/test_run.py:
```diff
@@ -1709,59 +1693,25 @@ async def main():
-                pass
+        with ignore_coroutine_never_awaited_warnings():
+            with pytest.raises(TypeError, match="expecting an async function"):
 
```
No need for an extra blank line after `with pytest.raises`.

oremanj replied:
You marked this as resolved but didn't change anything in your PR. Do you feel like the extra blank lines here are important for readability in a way that's related to the changes you're making here? It seems like you've added a number of them in this diff after block introducers (for, with, if, etc). That's not consistent with the style in the rest of the codebase.

oremanj commented on trio/_threads.py:
```diff
@@ -427,16 +427,18 @@ def from_thread_run_sync(fn, *args, trio_token=None):
-            if inspect.iscoroutine(call):
-                call.close()
+            if inspect.iscoroutine(ret):
+
```
See comment above about extra blank lines after block introducers

guilledk replied:
Ups sorry about that, it's burned into my brain

guilledk commented:
Ok now I think I got all the formatting issues, sorry guys very used to my code style. Thanks again for the patience.

oremanj commented:
Thank you for *your* patience! I removed a few more of the extraneous blank lines, and will merge this once the CI finishes.

oremanj commented:
Also, since I made some trivial commits above, you'll want to pull from your PR branch before you make additional changes. Sorry for the trouble!

guilledk commented:
No trouble at all! Now I understand the workflow a lot more, next time I'll take less reviews I swear haha

oremanj commented on newsfragments/1244.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Added helpful error message in the case the user passed incorrect type of function to "from_thread_run" & "from_thread_run_sync" functions.
```
```suggestion
Added a helpful error message if a sync function is passed to `trio.from_thread.run_sync` or an async function to `trio.from_thread.run`.
```

Reworded for clarity and to link to the documentation of those functions (that's what the single backticks do).

pquentin commented:
Congratulations @guilledk! How was the experience? Would you like to work on another issue next? https://github.com/python-trio/trio/issues/1532, for example?

guilledk commented:
Thanks @pquentin!, Awsome experience, to be honest never wrote a test before this PR, now all my projects have tests, so I learned a thing already.

I was actually looking at #1532 to do next, I really like the `trio` community and technology, I'd like to keep learning about the backend and helping you guys. Right now I have project that runs on top of `trio` and might go into production soon so I'd like to give back!

pquentin commented:
@guilledk Thanks, that's great to hear! If you decide to work on #1532, please ask if you have any questions, that might avoid a round of feedback or two. :)
